File: core/src/com/inkostilation/pong/engine/Score.java
package com.inkostilation.pong.engine;

import com.inkostilation.pong.commands.AbstractResponseCommand;
import com.inkostilation.pong.commands.response.ResponseScoreCommand;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Score implements ICommandSender {

    private final int maxScoreValue = 5;
    private Map<PlayerRole, Integer> points = new HashMap<>();
    private boolean changed = false;

    public Score() {
        points.put(PlayerRole.FIRST, 0);
        points.put(PlayerRole.SECOND, 0);
    }

    public int getPlayerScore(PlayerRole role) {
        return points.get(role);
    }

    public int getMaxScoreValue() {
        return maxScoreValue;
    }

    public void addPlayerScore(PlayerRole role, int add)
    {
        int val = points.get(role);
        points.replace(role, val + add);

        if (add != 0) {
            changed = true;
        }
    }

    public long getMaxValueCount() {
        return points.values()
                .stream()
                .filter(o -> o >= getMaxScoreValue())
                .count();
    }

    public List<PlayerRole> getMaxedPlayers() {
        return points.entrySet()
                .stream()
                .filter(o -> o.getValue() >= getMaxScoreValue())
                .map(o -> o.getKey())
                .collect(Collectors.toList());
    }

    @Override
    public boolean hasCommands() {
        return changed;
    }

    @Override
    public void getCommands(List<AbstractResponseCommand> pool) {
        if (changed) {
            pool.add(new ResponseScoreCommand(this));
            changed = false;
        }
    }
}

File: core/src/com/inkostilation/pong/engine/Score.kt
package com.inkostilation.pong.engine

import com.inkostilation.pong.commands.AbstractResponseCommand
import com.inkostilation.pong.commands.response.ResponseScoreCommand
import java.util.*
import java.util.stream.Collectors

class Score : ICommandSender {
    val maxScoreValue = 5
    private val points: MutableMap<PlayerRole, Int> = HashMap()
    private var changed = false
    fun getPlayerScore(role: PlayerRole?): Int {
        return points.get(role)
    }

    fun addPlayerScore(role: PlayerRole, add: Int) {
        val `val` = points[role]!!
        points.replace(role, `val` + add)
        if (add != 0) {
            changed = true
        }
    }

    val maxValueCount: Long
        get() = points.values
                .stream()
                .filter { o: Int -> o >= maxScoreValue }
                .count()

    val maxedPlayers: List<PlayerRole?>
        get() = points.entries
                .stream()
                .filter { o: Map.Entry<PlayerRole, Int> -> o.value >= maxScoreValue }
                .map { o: Map.Entry<PlayerRole, Int> -> o.key }
                .collect(Collectors.toList())

    override fun hasCommands(): Boolean {
        return changed
    }

    override fun getCommands(pool: MutableList<AbstractResponseCommand?>) {
        if (changed) {
            pool.add(ResponseScoreCommand(this))
            changed = false
        }
    }

    init {
        points[PlayerRole.FIRST] = 0
        points[PlayerRole.SECOND] = 0
    }
}